protofsm/state_machine.go

```diff
@@ -0,0 +1,428 @@
+)
+
+// EmitedEvent is a special type that can be emitted by a state transition.
+// This can container internal events which are to be routed back to the state,
+// or external events which are to be sent to the daemon.
+type EmittedEvent[Event any] struct {
+	// InternalEvent is an optional internal event that is to be routed
+	// back to the target state. This enables state to trigger one or many
+	// state transitions without a new external event.
+	InternalEvent fn.Option[Event]
+
+	// ExternalEvent is an optional external event that is to be sent to
+	// the daemon for dispatch. Usually, this is some form of I/O.
+	ExternalEvents fn.Option[DaemonEventSet]
+}
+
+// StateTransition is a state transition type. It denotes the next state to go
+// to, and also the set of events to emit.
+type StateTransition[Event any, Env Environment] struct {
+	// NextState is the next state to transition to.
+	NextState State[Event, Env]
+
+	// NewEvents is the set of events to emit.
+	NewEvents fn.Option[EmittedEvent[Event]]
+}
+
+// Environment is an abstract interface that represents the environment that
+// the state machine will execute using. From the PoV of the main state machine
+// executor, we just care about being able to clean up any resources that were
+// allocated by the environment.
+type Environment interface {
+	// CleanUp is a method that'll be called once the state machine has
+	// reached a terminal state. It signals the end of the execution of the
+	// state machine.
+	CleanUp() error
+
+	// TODO(roasbeef): also add checkpointing?
+}
+
+// State defines an abstract state along, namely its state transition function
+// that takes as input an event and an environment, and returns a state
+// transition (next state, and set of events to emit). As state can also either
+// be terminal, or not, a terminal event causes state execution to halt.
+type State[Event any, Env Environment] interface {
+	// ProcessEvent takes an event and an environment, and returns a new
+	// state transition. This will be iteratively called until either a
+	// terminal state is reached, or no further internal events are
+	// emitted.
+	ProcessEvent(event Event, env Env) (*StateTransition[Event, Env], error)
+
+	// IsTerminal returns true if this state is terminal, and false otherwise.
+	IsTerminal() bool
+
+	// TODO(roasbeef): also add state serialization?
+}
+
+// DaemonAdapters is a set of methods that server as adapters to bridge the
+// pure world of the FSM to the real world of the daemon. These will be used to
+// do things like broadcast transactions, or send messages to peers.
+type DaemonAdapters interface {
+	// SendMessages sends the target set of messages to the target peer.
+	SendMessages(btcec.PublicKey, []lnwire.Message) error
+
+	// BroadcastTransaction broadcasts a transaction with the target label.
+	BroadcastTransaction(*wire.MsgTx, string) error
+
+	// DisableChannel disables the target channel.
+	DisableChannel(wire.OutPoint) error
```

Comment:
Should we have `Enable` too?

Reply:
I think so, but not used by anything yet, so perhaps we can add when the first FSM needs it? One I can think of is going from the `stfu` state back to the normal one.